Bitcoin processor

vrx00 opened the pull request:
I completed and tested most of the part for bitcoin processor. I will keep improving it.

vrx00 committed:
serai Dockerfile & Makefile fixed

vrx00 committed:
Merge remote-tracking branch 'upstream/develop' into develop

vrx00 committed:
Merge remote-tracking branch 'upstream/develop' into develop

vrx00 committed:
Merge remote-tracking branch 'upstream/develop' into develop

vrx00 committed:
added new bitcoin mod & bitcoinhram

vrx00 committed:
couple changes

vrx00 committed:
merged develop branch

vrx00 committed:
merged from develop branch

vrx00 committed:
added odd&even check for bitcoin signing

vrx00 committed:
sign message updated

vrx00 committed:
print_keys commented out

vrx00 committed:
fixed signing process

vrx00 committed:
Merge remote-tracking branch 'upstream/develop' into bitcoin-processor

vrx00 committed:
Added new bitcoin library & added most of bitcoin processor logic

vrx00 committed:
added new crate and refactored the bitcoin coin library

vrx00 committed:
added signing test function

vrx00 committed:
moved signature.rs

vrx00 committed:
publish set to false

vrx00 committed:
Merge branch 'bitcoin-processor' of https://github.com/vrx00/serai into bitcoin-processor

vrx00 committed:
tests moved back to the root

vrx00 committed:
Merge remote-tracking branch 'upstream/develop' into bitcoin-processor

vrx00 committed:
added new functions to rpc

vrx00 committed:
added utxo test

vrx00 committed:
added new rpc methods and refactored bitcoin processor

vrx00 committed:
Merge remote-tracking branch 'upstream/develop' into bitcoin-processor

vrx00 committed:
added spendable output & fixed errors & added new logic for sighash & opened port 18443 for bitcoin docker

vrx00 committed:
changed tweak keys

vrx00 committed:
added tweak_keys & publish transaction and refactored bitcoin processor

vrx00 committed:
added new structs and fixed problems for testing purposes

vrx00 committed:
reverted dockerfile back its original

vrx00 committed:
reverted block generation of bitcoin to 5 seconds

vrx00 committed:
deleted unnecessary test function

vrx00 committed:
added new sighash & added new dbg messages & fixed couple errors

vrx00 committed:
fixed couple issue & removed unused functions

vrx00 committed:
fix for signing process

vrx00 committed:
crypto file for bitcoin refactored

vrx00 committed:
disabled test_send & removed some of the debug logs

vrx00 committed:
signing implemented & transaction weight calculation added & change address logic added

vrx00 committed:
refactored tweak_keys

vrx00 committed:
refactored mine_block & fixed change_address logic

vrx00 committed:
Merge remote-tracking branch 'upstream/develop' into bitcoin-processor

vrx00 committed:
implemented new traits to bitcoin processor& refactored bitcoin processor

vrx00 committed:
added new line to tests file

vrx00 committed:
added new line to bitcoin's wallet.rs

kayabaNerve commented on coins/Cargo.toml:
```diff
@@ -0,0 +1,8 @@
+[package]
```
This shouldn't exist. `coins/` is a folder, not a crate.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+[package]
+name = "bitcoin-serai"
+version = "0.1.0"
+edition = "2021"
```
Missing description, license (same as `coins/ethereum`, AGPL-3.0-only), and a copy of the AGPL short text as `LICENSE` (see `coins/ethereum` for an instance), repository, and authors.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+dalek-ff-group = { path = "../../crypto/dalek-ff-group", version = "0.1" }
+k256 = { version = "0.11", features = ["arithmetic", "keccak256", "ecdsa"] }
+secp256k1 = {version = "0.24.1", features = ["global-context", "rand-std"] }
+zeroize = { version = "1.5", features = ["zeroize_derive"] }
```
Missing new line.

vrx00 replied:
added new line

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+
+[dependencies]
+async-trait = "0.1.36"
+bitcoin = {version = "0.29.2", features=["serde"]}
```
Missing spaces, don't specify patch on this nor the package above it.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+bitcoin_hashes = "0.11.0"
+bitcoincore-rpc-json = "0.16.0"
+rand_core = "0.6"
+serde = {version = "1.0", features = ["derive"]}
```
Missing spaces on {}.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+anyhow = "1.0.32"
+tokio = { version = "1", features = ["full"] }
+thiserror = "1"
+sha2 = { version = "0.10" }
```
Unnecessary object since it's just the version field.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+tokio = { version = "1", features = ["full"] }
+thiserror = "1"
+sha2 = { version = "0.10" }
+frost = { version= "0.5", package = "modular-frost", path = "../../crypto/frost", features = ["secp256k1"] }
```
Missing space on `version=`.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+thiserror = "1"
+sha2 = { version = "0.10" }
+frost = { version= "0.5", package = "modular-frost", path = "../../crypto/frost", features = ["secp256k1"] }
+dleq = {  version = "0.2", package = "dleq", path = "../../crypto/dleq", features = ["serialize"]}
```
Extra space before version, unnecessary package name specification.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+name = "bitcoin-serai"
+version = "0.1.0"
+edition = "2021"
+publish = false
```
We likely do want to publish this as we do `coins/monero`, just to offer FROST utils around `bitcoin`. I'd like to keep this AGPL-3.0-only/`publish = false` for now yet come back to that.

vrx00 replied:
Makes sense. For now, I am keeping it as it is then.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+transcript = { package = "flexible-transcript", path = "../../crypto/transcript", version = "0.2", features = ["recommended"] }
+dalek-ff-group = { path = "../../crypto/dalek-ff-group", version = "0.1" }
+k256 = { version = "0.11", features = ["arithmetic", "keccak256", "ecdsa"] }
+secp256k1 = {version = "0.24.1", features = ["global-context", "rand-std"] }
```
Don't specify the patch version. cargo will grab the latest applicable subversion. Specifying `.1` as a subversion prevents us from using the latest subversion of `0.24`, missing bug fixes, and also adding two distinct versions of this crate to our depends.

Also, missing space before `version`.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+serde_derive = "1.0"
+log = "0.4.8"
+reqwest = { version = "0.11", features = ["json"] }
+anyhow = "1.0.32"
```
What's the justification for anyhow? `coins/ethereum` added some error thing because `ethers-rs` used it, so it was effectively required to. anyhow is generally intended for binaries though, so I'm unsure why it's here.

Also, don't specify the patch version.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+serde = {version = "1.0", features = ["derive"]}
+serde_json = "1.0"
+serde_derive = "1.0"
+log = "0.4.8"
```
Don't specify patch.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+bitcoin = {version = "0.29.2", features=["serde"]}
+hex = "0.4"
+bitcoin_hashes = "0.11.0"
+bitcoincore-rpc-json = "0.16.0"
```
Don't specify patch on either of the above two packages.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+    data.update(tag_hash);
+    data.update(r_encoded_point.x().to_owned().unwrap());
+    data.update(a_encoded_point.x().to_owned().unwrap());
+    data.update(&m[..]);
```
This should be able to just be `m`.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+
+    let r_encoded_point = R.to_encoded_point(true);
+    let a_encoded_point = A.to_encoded_point(true);
+    let tag = b"BIP0340/challenge";
```
Should be a constant.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+    let r_encoded_point = R.to_encoded_point(true);
+    let a_encoded_point = A.to_encoded_point(true);
+    let tag = b"BIP0340/challenge";
+    let tag_hash = Sha256::digest(tag);
```
We can use a lazy_static for this so it's only calculated once. https://github.com/serai-dex/serai/blob/develop/coins/monero/src/lib.rs#L88-L90 for an example.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+    data.update(tag_hash);
+    data.update(tag_hash);
+    data.update(r_encoded_point.x().to_owned().unwrap());
+    data.update(a_encoded_point.x().to_owned().unwrap());
```
I'm unsure if these to_owned are needed, as this returns a reference the Digest trait should be able to handle. If it doesn't, I *think* `.unwrap().as_ref()` will work without explicitly copying it.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+  }
+}
+
+pub fn taproot_key_spend_signature_hash(
```
Ideally, this wouldn't be public. It should be internal to `coins/bitcoin` which just exposes the FROST API over PSTs.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+  psbt: &psbt::PartiallySignedTransaction,
+  input_index: usize,
+) -> Result<(bitcoin::util::taproot::TapSighashHash, SchnorrSighashType), SignerError> {
+  if input_index >= psbt.inputs.len() || input_index >= psbt.unsigned_tx.input.len() {
```
Right now, this is public, so this is needed since it looks like pbst unfortunately doesn't guarantee its consistency.

If we make this non-public, we only need to run this check whenever we create the machine and take ownership of the PBST. Then, since we guaranteed consistency at the start, and no longer expose a mutation API, we won't need such checks.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+    .sighash_type
+    .unwrap_or_else(|| SchnorrSighashType::Default.into())
+    .schnorr_hash_ty()
+    .map_err(|_| SignerError::InvalidSighash)?;
```
This will accept any sighash defined here: https://docs.rs/bitcoin/latest/bitcoin/util/sighash/enum.SchnorrSighashType.html

Do we consider this a safety issue?

Serai itself should solely use All, and explicitly use All. `coins/bitcoin`, if we want to publish it to offer a Rust lib for FROST BTC TXs, shouldn't be so opinionated. It also isn't an issue for Serai since we aren't sending PBSTs over the wire, letting a malicious node instead use None to steal funds. We'll independently create TXs using All.

I just wanted to get your comment. This does seem to be fitting for `coins/bitcoin`.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+    .unwrap_or_else(|| SchnorrSighashType::Default.into())
+    .schnorr_hash_ty()
+    .map_err(|_| SignerError::InvalidSighash)?;
+  let witness_utxos = (0..psbt.inputs.len()).map(|i| psbt.get_utxo_for(i)).collect::<Vec<_>>();
```
filter_map here and it'll automatically exclude None, voiding the need for all_witness_utxos and the following if statement.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+  let mut cache = sighash::SighashCache::new(&psbt.unsigned_tx);
+  let prevouts = if witness_utxos.iter().all(Option::is_some) {
+    all_witness_utxos.extend(witness_utxos.iter().filter_map(|x| x.as_ref()));
+    sighash::Prevouts::All(&all_witness_utxos)
```
This hard-codes to All despite explicitly supporting any sighash type above. This should either use the prior-defined type or we should explicitly only support `All` right now. We shouldn't have halfway code which introduces the potential for inconsistencies. I'm not asking to support all variants, I couldn't care less right now, just to be consistent in what we support.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+
+  let hash = cache.taproot_key_spend_signature_hash(input_index, &prevouts, sighash_type);
+
+  Ok((hash.unwrap(), sighash_type))
```
This shouldn't unwrap since it works off unvalidated input.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+use anyhow::Result;
+use serde::{de::DeserializeOwned};
+use std::{collections::HashMap, fmt::Debug, vec::Vec, str::FromStr};
```
Unnecessary brackets on serde, and unnecessary import of Vec which is auto-imported.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+use std::{collections::HashMap, fmt::Debug, vec::Vec, str::FromStr};
+
+use bitcoin::{
+    hashes::hex::FromHex, secp256k1::ecdsa::Signature, Address, Amount, EcdsaSighashType, OutPoint,
```
This file has a 4-space indent?

vrx00 replied:
changed it to 2-space indent

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+    PrivateKey, Transaction, BlockHash,
+};
+
+use crate::rpc_helper::{
```
Should this just use `rpc_helper::*`?

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+    RpcConnectionError, RpcParams, RpcResponse, GetBlockWithDetailResult
+};
+
+use bitcoincore_rpc_json::{
```
Also, should this just use `*`?

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+}
+
+impl Rpc {
+    pub fn new(url: String, username: String, userpass: String) -> anyhow::Result<Rpc> {
```
This doesn't accept an URL. It accepts an IP address or domain. We then create an http URL out of it.

Rename the variable to IP? Take in a single URL string which is expected to be pre-formatted? Then it also supports un-authed RPC/RPC over https without further edits, and you don't have to craft a URL.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+            .map_err(|_| anyhow::Error::new(RpcConnectionError::ParsingError))?;
+        //dbg!(&parsed_res.error);
+        match parsed_res.error {
+            None => Ok(parsed_res.result.unwrap()),
```
I'm unsure this unwrap is safe. Right now, a node can return no error yet no result, causing a panic. I'm unsure if an honest BTC node will but...

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+    }
+
+    pub async fn get_height(&self) -> anyhow::Result<usize> {
+        let info: usize = self.rpc_call::<usize>("getblockcount", &[]).await?;
```
This specifies `info: usize` yet also `rpc_call::<usize>`. Only one should be needed.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        }
+    }
+
+    pub async fn get_height(&self) -> anyhow::Result<usize> {
```
Could we call this `get_latest_block_number`, please? Assuming that whatever this returns is always a valid block number on the chain it's describing.

Also, in general, we should have our APIs be get_latest_block_hash and get_block_number(hash). It ensures determinism. This function does still have value, I just wanted to leave that comment.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn get_block_info(&self, block_hash: &str) -> anyhow::Result<GetBlockResult> {
```
Remove, since we only use it to get the block number and I'd like to minimize this file please.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn get_block_index(&self, block_hash: &str) -> anyhow::Result<usize> {
```
get_block_number

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(bitcoin::consensus::encode::deserialize(&bytes)?)
+    }
+
+    pub async fn get_block_with_transactions(&self, block_hash: &bitcoin::BlockHash) -> anyhow::Result<GetBlockWithDetailResult> {
```
What's the difference between this and `get_block`? Shouldn't the block always have its TXs, from a design perspective? Else, it's effectively just the header.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn get_spendable(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn get_o_indexes(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn lock_unspent(&self, outputs: &[OutPoint]) -> Result<bool> {
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn unlock_unspent(&self, outputs: &[OutPoint]) -> Result<bool> {
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn get_fee_per_byte(&self, conf_target: Option<usize>) -> anyhow::Result<u64> {
```
Remove. While `estimatesmartfee` appears deterministic, that assumes a consistent best block which we can't guarantee, both due to competitors and due to latency.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(tx)
+    }
+
+    pub async fn get_transactions(
```
Remove get_transaction and get_transactions for functions (such as the following get_raw_transactions) supporting `bitcoin`'s Transaction, a usable and portable type.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(transactions)
+    }
+
+    pub async fn get_raw_transactions(
```
Just call get_transactions.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(transactions)
+    }
+
+    pub async fn is_confirmed(&self, tx_hash: &str) -> anyhow::Result<bool> {
```
Remove. While this is suitably deterministic, it hardcodes the confirmation count to 10. The scanner won't need this and I'd rather not support this.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok((self.get_height().await?.saturating_sub(tx_block_number) + 1) >= 10)
+    }
+
+    pub async fn get_transaction_block_number(&self, tx_hash: &str) -> anyhow::Result<usize> {
```
Is this needed? If not, please remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn get_block_transactions(
```
Remove as it's not using `bitcoin::Transaction`.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(transactions)
+    }
+
+    pub async fn get_block_raw_transactions(
```
Why do we have get_raw_transaction, get_block_raw_transactions, get_block, get_block_with_transactions, AND get_raw_transactions? Can we *please* just consolidate these.

The processor expects/supports either:
- Block with TXs
- Block without TXs

The former is superior and overall more efficient. The only reason we support the latter is for a Monero consideration/API boundary. Can we just have `get_block`, returning a bitcoin type with all its TXs, please? If we technically need to define a block as a block hash, and from there have get_block_transactions, fine, I just want to minimize this file.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(bitcoin::consensus::encode::deserialize(&bytes)?)
+    }
+
+    pub async fn get_raw_transaction_hex(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        self.rpc_call::<String>("getrawtransaction", &args).await
+    }
+
+    pub async fn get_raw_transaction_info(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        self.rpc_call::<String>("createrawtransaction", &args).await
+    }
+
+    pub async fn create_raw_transaction(
```
Remove this and its hex variant, please.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(bitcoin::consensus::encode::deserialize(&bytes)?)
+    }
+
+    pub async fn fund_raw_transaction<R: RawTx>(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+    }
+
+    #[deprecated]
+    pub async fn sign_raw_transaction<R: RawTx>(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+            .await
+    }
+
+    pub async fn sign_raw_transaction_with_wallet<R: RawTx>(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+            .await
+    }
+
+    pub async fn sign_raw_transaction_with_key<R: RawTx>(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+            .await
+    }
+
+    pub async fn send_raw_transaction<R: RawTx>(&self, tx: R) -> Result<bitcoin::Txid>
```
Should RawTx be a generic?

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn send_raw_str_transaction(&self, raw_tx: String) -> Result<bitcoin::Txid>
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn verify_message(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        self.rpc_call::<bool>("verifymessage", &args).await
+    }
+
+    pub async fn get_new_address(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        self.rpc_call::<Address>("getnewaddress", &args).await
+    }
+
+    pub async fn generate_to_address(&self, nblocks: usize, address: &str) -> Result<Vec<String>> {
```
Label test.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn get_balance(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(Amount::from_btc(self.rpc_call("getbalance", &args).await?)?)
+    }
+
+    pub async fn decode_raw_transaction<R: RawTx>(&self, tx: R) -> Result<GetBlockResult>
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+            .await
+    }
+
+    pub async fn list_transactions(
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc.rs:
```diff
@@ -0,0 +1,538 @@
+        Ok(info)
+    }
+
+    pub async fn test_mempool_accept<R: RawTx>(
```
Do we need this beyond? What does it offer sendraw doesn't?

kayabaNerve commented on coins/bitcoin/src/rpc_helper.rs:
```diff
@@ -0,0 +1,316 @@
+
+#[derive(Clone, PartialEq, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GetBlockWithDetailResult {
```
Remove.

kayabaNerve commented on coins/bitcoin/src/rpc_helper.rs:
```diff
@@ -0,0 +1,316 @@
+}
+
+#[derive(Deserialize, Debug)]
+pub struct RpcResponse<T> {
```
I believe if this is written as an enum, we can ensure it's either result or error, not both, resolving a prior comment.

Should support Clone.

kayabaNerve commented on coins/bitcoin/src/rpc_helper.rs:
```diff
@@ -0,0 +1,316 @@
+    pub error: Option<RpcResponseError>,
+}
+
+#[derive(Serialize)]
```
Why does this have a lifetime? Have it either take static strings or String.

kayabaNerve commented on coins/bitcoin/src/rpc_helper.rs:
```diff
@@ -0,0 +1,316 @@
+
+/// The error type for errors produced in this library.
+#[derive(Debug)]
+pub enum RpcError {
```
Derive Clone, PartialEq, Eq.

Use thiserror.

kayabaNerve commented on coins/bitcoin/src/rpc_helper.rs:
```diff
@@ -0,0 +1,316 @@
+    }
+}
+
+impl fmt::Display for RpcError {
```
Not needed with thiserror.

kayabaNerve commented on coins/bitcoin/src/rpc_helper.rs:
```diff
@@ -0,0 +1,316 @@
+    }
+}
+
+impl error::Error for RpcError {
```
Not needed with thiserror.

kayabaNerve commented on coins/bitcoin/src/rpc_helper.rs:
```diff
@@ -0,0 +1,316 @@
+    }
+}
+
+pub trait RawTx: Sized + Clone {
```
Can we just support Transaction and Transaction alone?

kayabaNerve reviewed (changes requested):
I only got through to the RPC and will review the rest later.

The RPC needs to be slashed and burned. Not only are several functions unusable to us provided, yet several inefficient/worse functions are provided when the optimal ones are right there. I've called out several functions for removal. That file should be rewritten as cleanly and minimally as possible. Only functions we are using or explicitly plan to use.

A lot of this is extremely basic. I'd imagine it take 30-60m to work through this as it's generally formatting/misc comments. Just needs to be cleaned up.

Happy to see the PR. Will finish review later.

kayabaNerve commented on coins/bitcoin/src/rpc_helper.rs:
```diff
@@ -0,0 +1,316 @@
+}
+
+/// Shorthand for converting an Option into an Option<serde_json::Value>.
+pub fn opt_into_json<T>(opt: Option<T>) -> Result<serde_json::Value>
```
These functions shouldn't be `pub`. Additionally, I think serde_json already converts None to null.

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+use bitcoin::{
+  util::{
+    schnorr::{SchnorrSig},
+    sighash::{SchnorrSighashType},
```
Unnecessary {}s.

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+use bitcoin_hashes::hex::ToHex;
+use frost::{
+  algorithm::Schnorr,
+  curve::{Secp256k1},
```
{}

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+  },
+};
+use crate::crypto::{BitcoinHram, make_even, taproot_key_spend_signature_hash};
+use rand_core::{RngCore};
```
{}

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+use crate::crypto::{BitcoinHram, make_even, taproot_key_spend_signature_hash};
+use rand_core::{RngCore};
+
+use core::{fmt::Debug};
```
{}

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+    self,
+    keys: ThresholdKeys<Secp256k1>,
+    mut transcript: RecommendedTranscript,
+    height: usize,
```
Let's try not use "height" as it's ambiguous between currencies. This should be the latest block's "number".

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+    let mut sigs = vec![];
+    let algorithm = Schnorr::<Secp256k1, BitcoinHram>::new();
+    for (i, input) in raw_tx.input.iter().enumerate() {
+      let txid: [u8; 32] = input.previous_output.txid.to_vec()[0..32].try_into().unwrap();
```
Is the TX ID ever not 32 bytes? If it's 32 bytes, you can don't need the `[.. 32]`. Also, you shouldn't need the .to_vec here.

(so ideally just `input.previous_output.txid.try_into().unwrap()`, yet you may need `input.previous_output.txid.as_ref().try_into().unwrap()`)

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+      transcript.append_message(b"input_hash", txid);
+      transcript.append_message(b"input_output_index", input.previous_output.vout.to_le_bytes());
+      transcript.append_message(
+        b"input_shared_key",
```
"shared_key"? The following variable is "internal_key". I'm unsure what the terminology is here.

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+        self.tx.inputs[i].tap_internal_key.unwrap().serialize(),
+      );
+
+      sigs.push(AlgorithmMachine::new(algorithm.clone(), keys.clone()).unwrap());
```
We have to tweak the keys to be even, and then further tweak per-input, right? The 'key path'? Shouldn't that occur here?

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+
+    for payment in raw_tx.output {
+      //TODO: need address generator function to generate address from script_pubkey
+      transcript.append_message(b"payment_address", payment.script_pubkey.as_bytes());
```
Payment is intended to be (Address, amount). It's perfectly fine to transcript `output_script` and `output_amount` here though. This doesn't need to operate off "payments", which is a universal format.

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+    total_weight += total_outputs * 8 * 4;
+    total_weight += total_outputs * 1 * 4;
+    // Script pubkey
+    total_weight += (address.script_pubkey().to_hex().len()/2) * 1 * 4;
```
as_bytes, don't use hex

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+    return Ok(TransactionMachine { signable: self, transcript, sigs });
+  }
+
+  pub fn calculate_weight(total_inputs: usize, address: &bitcoin::Address, change: bool) -> usize{
```
Can't we use a function from `bitcoin` OR calculate via `.serialize().len()`? I understand the latter is faulty regarding Segwit weight, but it'd be better to get the base cost than this.

Also, why does this take a single address out?

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+        "message was passed to the TransactionMachine when it generates its own",
+      ))?;
+    }
+    let mut msg_list = Vec::new();
```
We should calculate this at time of use, not here.

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+      .drain(..)
+      .enumerate()
+      .map(|(index, sig)| {
+        let (sig, share) = sig.sign(commitments.remove(0), &msg_list.remove(index))?;
```
This iterates indexes 0 to n, yet calls remove. This either needs to remove 0, like commitments, or not remove from the list yet just calculate the msg here. Right now, for two inputs, it'll remove 0, be left with what was 1 at position 0, then try to remove 1, failing as nothing is currently at 1.

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+  }
+}
+
+impl SignatureMachine<PartiallySignedTransaction> for TransactionSignatureMachine {
```
This should return a Transaction, not a PBST.

kayabaNerve commented on coins/bitcoin/src/transactions.rs:
```diff
@@ -0,0 +1,288 @@
+  type SignatureShare = Vec<SignatureShare<Secp256k1>>;
+
+  fn read_share<R: Read>(&self, reader: &mut R) -> io::Result<Self::SignatureShare> {
+    self.sigs.iter().map(|clsag| clsag.read_share(reader)).collect()
```
"clsag"

kayabaNerve commented on coins/bitcoin/src/wallet.rs:
```diff
@@ -0,0 +1,32 @@
+use bitcoin::{Txid};
```
{}

kayabaNerve commented on coins/bitcoin/src/wallet.rs:
```diff
@@ -0,0 +1,32 @@
+  pub fn serialize(&self) -> Vec<u8> {
+    let mut res = self.txid.to_vec();
+    res.extend(self.vout.to_le_bytes().to_vec());
+    res.extend(self.amount.to_le_bytes().to_vec());
```
Don't use to_vec just to extend off it. If the raw bytes don't work, just borrow with `&`.

kayabaNerve commented on coins/bitcoin/tests/signature.rs:
```diff
@@ -0,0 +1,47 @@
+  dbg!(pubkey_compressed);
+  
+  let algo = Schnorr::<Secp256k1, BitcoinHram>::new();
+  let mut _sig = sign(
```
Don't use `_sig` for a used variable. Use `sig`.

kayabaNerve commented on coins/bitcoin/tests/signature.rs:
```diff
@@ -0,0 +1,47 @@
+
+  let mut _offset = 0;
+  (_sig.R, _offset) = make_even(_sig.R);
+  _sig.s += Scalar::from(_offset);
```
Ideally, we'd have a BitcoinSchnorr Algorithm to handle this for us.

kayabaNerve commented on coins/bitcoin/tests/signature.rs:
```diff
@@ -0,0 +1,47 @@
+  let pubkey =
+    secp256k1::XOnlyPublicKey::from_slice(&pubkey_compressed.x().to_owned().unwrap()).unwrap();
+  dbg!(pubkey.to_hex().to_string());
+  let _res = sig.verify(&msg, &pubkey).unwrap();
```
Don't assign `_res`. `.unwrap()` is sufficient to handle the result.

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+dleq = {  version = "0.2", package = "dleq", path = "../../crypto/dleq", features = ["serialize"]}
+transcript = { package = "flexible-transcript", path = "../../crypto/transcript", version = "0.2", features = ["recommended"] }
+dalek-ff-group = { path = "../../crypto/dalek-ff-group", version = "0.1" }
+k256 = { version = "0.11", features = ["arithmetic", "keccak256", "ecdsa"] }
```
Don't add keccak nor ecdsa if not needed? Neither should be?

kayabaNerve commented on processor/Cargo.toml:
```diff
@@ -25,9 +25,20 @@ curve25519-dalek = { version = "3", features = ["std"] }
 monero-serai = { path = "../coins/monero", features = ["multisig"] }
+bitcoin-serai = { path = "../coins/bitcoin" }
+
+k256 = { version = "0.11", features = ["arithmetic", "keccak256", "ecdsa"] }
```
Same comment on keccak/ecdsa features.

kayabaNerve commented on processor/Cargo.toml:
```diff
@@ -25,9 +25,20 @@ curve25519-dalek = { version = "3", features = ["std"] }
+bitcoin-serai = { path = "../coins/bitcoin" }
+
+k256 = { version = "0.11", features = ["arithmetic", "keccak256", "ecdsa"] }
+sha2 = { version = "0.10" }
```
Unnecessary object.

kayabaNerve commented on processor/Cargo.toml:
```diff
@@ -25,9 +25,20 @@ curve25519-dalek = { version = "3", features = ["std"] }
+bitcoincore-rpc-json = "0.16.0"
+secp256k1 = {version = "0.24.1", features = ["global-context", "rand-std"] }
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
```
Please correct spacing and remove patch version specification.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+use async_trait::async_trait;
+use bitcoin_hashes::hex::{FromHex, ToHex};
+use std::{str::FromStr};
+use transcript::{RecommendedTranscript};
```
{}

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+use transcript::{RecommendedTranscript};
+
+use frost::{
+  curve::{Secp256k1},
```
{}

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+};
+
+use bitcoin::{
+  Block as MBlock,
```
BBlock

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+use bitcoin::{
+  Block as MBlock,
+  util::address::Address,
+  Txid, schnorr::{TweakedPublicKey},
```
{}

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+use bitcoin_serai::{
+  rpc::Rpc,
+  rpc_helper::RawTx,
+  crypto::{make_even},
```
{}

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+  rpc::Rpc,
+  rpc_helper::RawTx,
+  crypto::{make_even},
+  wallet::{SpendableOutput},
```
{}

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+  rpc_helper::RawTx,
+  crypto::{make_even},
+  wallet::{SpendableOutput},
+  transactions::{TransactionMachine, SignableTransaction as MSignableTransaction},
```
BSignable

kayabaNerve commented on coins/bitcoin/src/wallet.rs:
```diff
@@ -0,0 +1,32 @@
+
+#[derive(Clone, Debug)]
+pub struct SpendableOutput {
+  pub txid: Txid,//[u8;32],
```
This can be a Txid. It doesn't have to be a [u8; 32].

kayabaNerve commented on coins/bitcoin/src/wallet.rs:
```diff
@@ -0,0 +1,32 @@
+use bitcoin_hashes::hex::FromHex;
+
+#[derive(Clone, Debug)]
+pub struct SpendableOutput {
```
Can we not just wrap a bitcoin type for these fields? Like Vout? And use its ser/deser?

kayabaNerve replied:
OutPoint specifically could merge these top two fields.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+
+use k256::{
+  ProjectivePoint, Scalar,
+  elliptic_curve::{
```
{}

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub struct Fee {
+  pub per_weight: u64,
+  pub mask: u64,
```
Bitcoin doesn't have a mask

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+}
+impl Bitcoin {
+  pub async fn new(url: String, username: Option<String>, userpass: Option<String>) -> Bitcoin {
+    Bitcoin { rpc: Rpc::new(url, username.unwrap(), userpass.unwrap()).unwrap()}
```
Should we just use a pre-formatted URL string?

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+
+  fn id(&self) -> Self::Id {
+    let serialized_data = self.0.serialize();
+    let ret: [u8; 36] = serialized_data[0..36].try_into().unwrap();
```
This is dependent on a different crate's serialization format. Please redo the serialization for TX hash/vout index here.

kayabaNerve replied:
Moving to Outpoint would potentially help here.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+  type Curve = Secp256k1;
+
+  type Fee = Fee;
+  type Transaction = PartiallySignedTransaction;
```
This isn't correct. It should be bitcoin::Transaction.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+  async fn get_block(&self, number: usize) -> Result<Self::Block, CoinError> {
+    let block_hash = self.rpc.get_block_hash(number - 1).await.unwrap();
+    let info = self.rpc.get_block(&block_hash).await.unwrap();
+    let mblock = Block(info.block_hash().to_vec()[0..32].try_into().unwrap(),info);
```
You shouldn't need to_vec here.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+    let mut outputs = Vec::new();
+    for one_transaction in block_details.tx {
+      for output_tx in one_transaction.vout {
+        if output_tx.script_pub_key.script().unwrap().cmp(&main_addr.script_pubkey()).is_eq() {
```
You should just be able to use `==`

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+      });
+    }
+
+    let transaction_weight = MSignableTransaction::calculate_weight(vin_alt_list.len(), &payments[0].0, false);
```
This only uses a single payment to calculate weight.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+    let target_sat = actual_fee + payment_sat;
+    if input_sat < target_sat {
+      //TODO: Change it with TransactionError
+      return Err(CoinError::ConnectionError);
```
We should have an error in Coin, `NotEnoughFunds`. Sorry it wasn't already there.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+      let transaction_weight = MSignableTransaction::calculate_weight(vin_alt_list.len(), &payments[0].0, true);
+      actual_fee = fee.calculate(transaction_weight);
+      let target_sat = actual_fee + payment_sat;
+      if target_sat < input_sat {
```
This is incomplete. Bitcoin won't relay outputs worth less than it costs to spend them. I think the formula we want to use is *something* like `target_sat < (input_sat - (fee.calculate(calculate_weight(&[vin[0], &[payments[0]], false) * 2))`. Say the remaining change has to be able to fund 2x the fees for a 1:1 TX.

Additionally, at some point, we need to drop outputs worth less than that...

Feel free to mark this TODO. Just needs to be documented.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+  const MAX_OUTPUTS: usize = 16;
+
+  fn address(&self, key: ProjectivePoint) -> Self::Address {
+    assert!(key.to_encoded_point(true).tag() == Tag::CompressedEvenY, "YKey is odd");
```
debug_assert

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+      let xonly_pubkey = XOnlyPublicKey::from_slice(&keys.group_key().to_encoded_point(true).x().to_owned().unwrap()).unwrap();
+      psbt.inputs[i].witness_utxo = Some(one_transaction.output[one_input.0.vout as usize].clone());
+      psbt.inputs[i].sighash_type = Some(PsbtSighashType::from_u32(SchnorrSighashType::All as u32));
+      psbt.inputs[i].tap_internal_key = Some(xonly_pubkey);
```
Isn't the internal key supposed to tweaked per the key path? If this passes tests, it passes tests, but https://bitcoin.stackexchange.com/questions/110402/how-to-tweak-a-public-key-for-taproot says it shouldn't AFAICT. Feel free to comment if you know more.

kayabaNerve replied:
I have no idea how this passes, if it does. Congrats if so, not like I care to complain it works... Would still love to know why.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+      psbt.inputs[i].sighash_type = Some(PsbtSighashType::from_u32(SchnorrSighashType::All as u32));
+      psbt.inputs[i].tap_internal_key = Some(xonly_pubkey);
+    }
+    return Ok(SignableTransaction { keys: keys, transcript: transcript, height: block_number+1, actual: MSignableTransaction{tx: psbt, fee:actual_fee} });
```
Can we please not use height here? I'm pretty sure Bitcoin treats height as the latest block number, making this incorrect.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+    tx: &Self::Transaction,
+  ) -> Result<(Vec<u8>, Vec<<Self::Output as OutputTrait>::Id>), CoinError> {
+    let target_tx = tx.clone().extract_tx();
+    let s_raw_transaction = self.rpc.send_raw_str_transaction(target_tx.raw_hex()).await.unwrap();
```
Why is this in hex? There should be a bytes option.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+  async fn publish_transaction(
+    &self,
+    tx: &Self::Transaction,
+  ) -> Result<(Vec<u8>, Vec<<Self::Output as OutputTrait>::Id>), CoinError> {
```
Coin shouldn't return Output IDs. You matched Monero here, and the API Coin defined, yet as I'm reading this, it makes no sense.... I can edit develop or you're welcome to add it to this PR.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+      .enumerate()
+      .map(|(i, output)| {
+        let one_output = SpendableOutput {
+          txid: Txid::from_str(target_tx.txid().to_string().as_str()).unwrap(),
```
You get a TX ID, convert it to a string, just to deserialize it from a string.
`target_tx.txid()`

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+    Ok((s_raw_transaction.to_vec(), vec_output))
+  }
+
+  fn tweak_keys<'a>(&self, key: &'a mut ThresholdKeys<Self::Curve>) {
```
This shouldn't need a lifetime.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+
+  #[cfg(test)]
+  async fn mine_block(&self) {
+    use bitcoincore_rpc_json::{AddressType};
```
{}

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+  async fn mine_block(&self) {
+    use bitcoincore_rpc_json::{AddressType};
+
+    let new_addr = self.rpc.get_new_address(None, Some(AddressType::Bech32)).await.unwrap();
```
Can't we just generate a garbage address using the bitcoin lib so we can purge reliance on the often-changing bitcoin wallet RPC?

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+  }
+
+  #[cfg(test)]
+  async fn test_send(&self, address: Self::Address) {
```
On the one hand, I'm fascinated by this using FROST itself. On the other hand, this should just send a small amount of funds to the provided address. Just generate a P2PKH with `bitcoin` you receive to (using `gen_to_addr`), and send from there. That way, if scanning breaks, we don't need to wonder if test_send messed up or if scanning did.

kayabaNerve commented on processor/src/coin/mod.rs:
```diff
@@ -106,4 +111,5 @@ pub trait Coin {
 
   #[cfg(test)]
   async fn test_send(&self, key: Self::Address);
+
```
Random new line?

kayabaNerve commented on processor/src/coin/monero.rs:
```diff
@@ -1,3 +1,5 @@
+use std::collections::HashMap;
```
Not needed import.

kayabaNerve commented on processor/src/coin/mod.rs:
```diff
@@ -98,6 +101,8 @@ pub trait Coin {
     tx: &Self::Transaction,
   ) -> Result<(Vec<u8>, Vec<<Self::Output as Output>::Id>), CoinError>;
 
+  fn tweak_keys<'a>(&self, key: &'a mut ThresholdKeys<Self::Curve>);
```
Shouldn't need lifetime.

kayabaNerve commented on processor/src/tests/mod.rs:
```diff
@@ -61,9 +61,19 @@ impl Network for LocalNetwork {
+
+  let xkey = keys[&1].group_key();
+  coin.address(xkey);
+  coin.mine_block().await;
```
I think these three lines are old code, no longer relevant.

kayabaNerve commented on processor/src/tests/mod.rs:
```diff
@@ -93,16 +103,27 @@ async fn test_send<C: Coin + Clone>(coin: C, fee: C::Fee) {
-    let signable = wallet
-      .prepare_sends(1, vec![(wallet.address(), 10000000000)], fee)
+    let mut signable = wallet
+      .prepare_sends(1, vec![(wallet.address(), 4999900000)], fee)
```
100000000? Keep it round?

kayabaNerve commented on processor/src/tests/mod.rs:
```diff
@@ -93,16 +103,27 @@ async fn test_send<C: Coin + Clone>(coin: C, fee: C::Fee) {
-    futures.push(wallet.attempt_send(network, signable));
+      .unwrap();
+    if signable.1.len() <= 0 {
+      break;
```
This shouldn't break here. That invalidates this test by silently returning if it fails.

kayabaNerve commented on processor/src/wallet.rs:
```diff
@@ -254,14 +254,12 @@ impl<D: CoinDb, C: Coin> Wallet<D, C> {
-
       let block = self.coin.get_block(b).await?;
       for (keys, outputs) in self.keys.iter_mut() {
+        self.coin.tweak_keys(keys);
```
Except for this one line, I don't believe any lines here are actually needed. `git checkout -f processor/src/wallet.rs`?

We should tweak keys at time of addition to self.keys so we only tweak them once.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+        let one_output = SpendableOutput {
+          txid: Txid::from_str(target_tx.txid().to_string().as_str()).unwrap(),
+          amount: output.value,
+          vout: i as u32,
```
Please use the explicit `u32::try_from(i).unwrap()`.

kayabaNerve commented on coins/bitcoin/src/crypto.rs:
```diff
@@ -0,0 +1,135 @@
+      if let Some(wit_utxo) = &input.witness_utxo {
+        Some(wit_utxo.clone())
+      } else if let Some(in_tx) = &input.non_witness_utxo {
+        Some(in_tx.output[tx.input[input_index].previous_output.vout as usize].clone())
```
Please use the explicit `usize::try_from(tx.input[input_index].previous_output.vout).unwrap()`.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+  async fn get_latest_block_number(&self) -> Result<usize, CoinError> {
+    let block_result = self.rpc.get_height().await.map_err(|_| CoinError::ConnectionError);
+    let block_number = match block_result {
+      Ok(val) => Ok(val as usize),
```
Please use explicit casts here, if needed. This RPC function should just return usize from the get go...

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+
+  async fn get_latest_block_number(&self) -> Result<usize, CoinError> {
+    let block_result = self.rpc.get_height().await.map_err(|_| CoinError::ConnectionError);
+    let block_number = match block_result {
```
Use `?` on the prior line so you don't have to match here.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,376 @@
+      let one_transaction = self.rpc.get_raw_transaction(&one_input.0.txid, None, None).await.unwrap();
+      let xonly_pubkey = XOnlyPublicKey::from_slice(&keys.group_key().to_encoded_point(true).x().to_owned().unwrap()).unwrap();
+      psbt.inputs[i].witness_utxo = Some(one_transaction.output[one_input.0.vout as usize].clone());
+      psbt.inputs[i].sighash_type = Some(PsbtSighashType::from_u32(SchnorrSighashType::All as u32));
```
Explicit casts. For the latter, `SchnorrSighashType::All.into()` should work. Else, `PsbtSighashType::from(SchnorrSighashType::All)` definitely will.

kayabaNerve commented:
Finished the review. A lot of these comments are duplicates/brief. We can also make some topics, like a BitcoinSchnorr algorithm, fee code, etc TODOs. It's just important to start getting things tracked.

Please merge develop in and run `cargo +nightly fmt` to get things started. Then I'd resolve clippy (also nightly), then start making changes.

kayabaNerve commented:
If clippy complains about elliptic curve variables being upper case, add `#[allow(clippy::non_snake_case)]` or whatever the label is to the line in question.

vrx00 committed:
deleted Cargo.toml from coins folder

vrx00 committed:
edited bitcoin's Cargo.toml and added LICENSE

vrx00 committed:
added new line to bitcoin's Cargo.toml

vrx00 committed:
added spaces

vrx00 committed:
added spaces

vrx00 committed:
deleted unnecessary object

vrx00 committed:
added spaces

vrx00 committed:
deleted patch numbers

vrx00 committed:
updated sha256 parameter for message

vrx00 committed:
updated tag as const

vrx00 committed:
deleted unnecessary brackets and imports

vrx00 committed:
updated rpc.rs to 2 space indent

vrx00 committed:
deleted unnecessary brackers

vrx00 committed:
deleted unnecessary brackets

vrx00 committed:
changed it to explicit

vrx00 committed:
updated to explicit

vrx00 committed:
deleted unnecessary parsing

vrx00 committed:
added ? for easy return

vrx00 committed:
updated imports

vrx00 committed:
updated height to number

vrx00 committed:
deleted unnecessary brackets

vrx00 committed:
updated clsag to sig & to_vec to as_ref

vrx00 committed:
updated _sig to schnorr_signature

vrx00 committed:
deleted unnecessary variable

vrx00 committed:
updated Cargo.toml of processor and bitcoin

vrx00 committed:
updated imports of bitcoin processor

vrx00 committed:
updated MBlock to BBlock

vrx00 committed:
updated MSignable to BSignable

vrx00 committed:
updated imports

vrx00 committed:
deleted mask from Fee

vrx00 committed:
updated get_block function return

vrx00 committed:
updated comparison logic for scripts

vrx00 committed:
updated assert to debug_assert

vrx00 committed:
updated height to number

vrx00 committed:
updated txid logic

vrx00 committed:
updated tweak_keys definition

vrx00 committed:
updated imports

vrx00 committed:
deleted new line

vrx00 committed:
delete HashMap from monero

vrx00 committed:
deleted old test code parts

vrx00 committed:
updated test amount to a round number

vrx00 committed:
changed the test code part back to its original

vrx00 committed:
updated imports of rpc.rs

vrx00 committed:
deleted unnecessary return assignments

vrx00 committed:
deleted get_fee_per_byte

vrx00 committed:
deleted create_raw_transaction

vrx00 committed:
deleted fund_raw_transaction

vrx00 committed:
deleted sign transaction rpc

vrx00 committed:
delete verify_message rpc

vrx00 committed:
deleted get_balance

vrx00 committed:
deleted decode_raw_transaction rpc

vrx00 committed:
deleted list_transactions rpc

vrx00 committed:
changed test_send to p2wpkh

vrx00 committed:
updated imports of test_send

vrx00 committed:
fixed imports of test_send

vrx00 committed:
updated bitcoin's mine_block function

vrx00 committed:
updated bitcoin's test_send

vrx00 committed:
updated bitcoin's hram and test_signing

vrx00 committed:
deleted 2 rpc function (is_confirmed & get_transaction_block_number)

vrx00 committed:
deleted get_raw_transaction_hex

vrx00 committed:
deleted get_raw_transaction_info

vrx00 committed:
deleted new_address

vrx00 committed:
deleted test_mempool_accept

vrx00 committed:
updated remove(0) to remove(index)

vrx00 committed:
deleted ger_raw_transaction

vrx00 committed:
deleted RawTx trait and converted type to Transaction

vrx00 committed:
reverted raw_hex feature back

vrx00 committed:
added NotEnoughFunds to CoinError

vrx00 committed:
changed Sighash to all

vrx00 committed:
removed lifetime of RpcParams

vrx00 committed:
changed pub to pub(crate) & changed sig_hash line

vrx00 committed:
changed taproot_key_spend_signature_hash to internal

vrx00 committed:
added Clone to RpcError & deleted get_utxo_for

vrx00 committed:
changed to_hex to as_bytes for weight calculation

vrx00 committed:
updated SpendableOutput

vrx00 committed:
deleted unnecessary parentheses

vrx00 committed:
updated serialize of Output s id field

vrx00 committed:
deleted unused crate & added lazy_static

vrx00 committed:
updated RPC init function

vrx00 committed:
added lazy_static for TAG_HASH & updated imported crates

vrx00 committed:
changed get_block_index to get_block_number

vrx00 committed:
deleted get_block_info

vrx00 committed:
updated get_height to get_latest_block_number

vrx00 committed:
removed GetBlockWithDetailResult and get_block_with_transactions

vrx00 committed:
deleted unnecessary imports from rpc_helper

vrx00 committed:
removed lock and unlock_unspent

vrx00 committed:
deleted get_transactions and get_transaction and renamed get_raw_transaction to get_transaction

vrx00 committed:
updated opt_into_json

vrx00 committed:
changed payment_address and amount to output_script and amount for transcript

vrx00 committed:
refactored error logic for rpc & deleted anyhow crate

vrx00 committed:
added a dedicated file for json helper functions

vrx00 committed:
refactored imports and deleted unused code

vrx00 committed:
added clippy::non_snake_case

vrx00 committed:
removed unused Error items

kayabaNerve commented on coins/bitcoin/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+dalek-ff-group = { path = "../../crypto/dalek-ff-group", version = "0.1" }
+k256 = { version = "0.11", features = ["arithmetic"] }
+secp256k1 = {version = "0.24", features = ["global-context", "rand-std"] }
+lazy_static = "1"
```
Missing newline.

kayabaNerve commented on coins/bitcoin/src/lib.rs:
```diff
@@ -0,0 +1,6 @@
+pub mod crypto;
+pub mod wallet;
+pub mod transactions;
+pub mod json_helper;
```
Missing newline.

kayabaNerve commented on coins/bitcoin/src/json_helper.rs:
```diff
@@ -0,0 +1,62 @@
+    Ok(into_json(Some(opt))?)
+}
+
+pub trait RawTx: Sized + Clone {
```
Why keep this around? We only have `Transaction` now, as we should.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,380 @@
+    let public_key = PublicKey::from_private_key(&secp, &private_key);
+    
+    let new_addr = Address::p2wpkh(&public_key, Network::Regtest).unwrap();
+    self.rpc.generate_to_address(1, new_addr.to_string().as_str()).await.unwrap();
```
You can use `&` to get a `&str`. `as_str()` isn't needed. (`&new_addr.to_string()`)

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,380 @@
+    let private_key = PrivateKey::new(secret_key, Network::Regtest);
+    let public_key = PublicKey::from_private_key(&secp, &private_key);
+    
+    let main_addr = Address::p2wpkh(&public_key, Network::Regtest).unwrap();
```
Create a dedicated test function to get a random key/address?

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,380 @@
+    let mut vout_list = Vec::new();
+
+    let new_block = self.get_latest_block_number().await.unwrap() + 1;
+    self.rpc.generate_to_address(1, main_addr.to_string().as_str()).await.unwrap();
```
Create a dedicated test function to get an unlocked coinbase output?

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,380 @@
+    };
+    
+    let mut sig_hasher = SighashCache::new(&new_transaction);
+    let script_code = Address::p2pkh(&public_key, Network::Testnet).script_pubkey();
```
Why Testnet when the rest used Regtest?

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,380 @@
+    signed_witness.push(public_key.to_bytes());
+    new_transaction.input[0].witness = signed_witness;
+
+    let _result = self.rpc.send_raw_transaction(&new_transaction).await.unwrap();
```
`let _result =` shouldn't be needed.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,380 @@
+    let mut signed_witness = Witness::new();
+    signed_witness.push_bitcoin_signature(&signed_der, EcdsaSighashType::All);
+    signed_witness.push(public_key.to_bytes());
+    new_transaction.input[0].witness = signed_witness;
```
This *could* be simplified by using P2PKH. I don't care to ask you to move to an inferior format, from a working piece of code, in the name of simplicity, yet I did want to note the thought.

kayabaNerve commented on processor/src/coin/mod.rs:
```diff
@@ -1,4 +1,4 @@
-use std::marker::Send;
+use std::{marker::Send,collections::HashMap};
```
Should no longer be necessary.

kayabaNerve commented on processor/src/tests/mod.rs:
```diff
@@ -96,13 +102,20 @@ async fn test_send<C: Coin + Clone>(coin: C, fee: C::Fee) {
     futures.push(wallet.attempt_send(network, signable));
   }
+  
+  if futures.len() > 0 {
```
This if shouldn't exist. There should always be a future.

It also somewhat-voids the test.

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,380 @@
+impl BlockTrait for Block {
+  type Id = [u8; 32];
+  fn id(&self) -> Self::Id {
+    self.0
```
self.0.block_hash()

kayabaNerve commented on processor/src/coin/bitcoin.rs:
```diff
@@ -0,0 +1,380 @@
+};
+
+#[derive(Clone, Debug)]
+pub struct Block([u8; 32], BBlock);
```
You don't need the `[u8; 32]` since you can get the hash from the block itself.

kayabaNerve replied:
Actually, you shouldn't even need the tuple. You should be able to implement BlockTrait directly on Block.

kayabaNerve reviewed (commented):
Few misc items I noted. Didn't do a full re-review.

vrx00 committed:
added new line to Cargo

vrx00 committed:
rekmoved Block and used bitcoin::Block direcetly

vrx00 committed:
removed added println and futures.len check

vrx00 committed:
removed HashMap from coin mod.rs

vrx00 committed:
updated Testnet to Regtest

vrx00 committed:
removed unnecessary variable

vrx00 committed:
updated as_str to &

vrx00 committed:
removed RawTx trait

vrx00 committed:
added newline

vrx00 committed:
changed test transaction to p2pkh

vrx00 committed:
updated test_send

vrx00 committed:
updated test_send

vrx00 committed:
updated test_send

vrx00 committed:
reformatted bitcoin processor

vrx00 committed:
moved sighash logic into signmachine

vrx00 committed:
removed generate_to_address

vrx00 committed:
added test_address function to bitcoin processor

vrx00 committed:
updated RpcResponse to enum and added Clone trait

vrx00 committed:
removed old RpcResponse

vrx00 committed:
updated shared_key to internal_key

vrx00 committed:
updated fee part

vrx00 committed:
updated test_send block logic

vrx00 committed:
added a test function for getting spendables

vrx00 committed:
updated tweaking keys logic

vrx00 committed:
updated calculate_weight logic

vrx00 committed:
added todo for BitcoinSchnorr Algorithm

vrx00 committed:
updated calculate_weight

vrx00 committed:
updated calculate_weight

vrx00 committed:
updated calculate_weight

vrx00 committed:
added a TODO for bitcoin's signing process

vrx00 committed:
removed unused code

kayabaNerve committed:
Merge branch 'develop' into bitcoin-processor

kayabaNerve committed:
Finish merging develop

kayabaNerve committed:
cargo fmt

kayabaNerve committed:
cargo machete

kayabaNerve committed:
Handle most clippy lints on bitcoin

Doesn't handle the unused transcript due to pending cryptographic considerations.

kayabaNerve committed:
Rearrange imports and clippy tests

kayabaNerve committed:
Misc processor lint

kayabaNerve committed:
Update deny.toml

kayabaNerve committed:
Remove unnecessary RPC code

vrx00 committed:
updated test_send

vrx00 committed:
added bitcoin ci & updated test-dependencies yml

vrx00 committed:
fixed bitcoin ci

kayabaNerve committed:
Remove mining from the bitcoin/monero docker files

The tests should control block production in order to test various
circumstances. The automatic mining disrupts assumptions made in testing. Since
we're now using the Bitcoin docker container for testing...

vrx00 committed:
updated bitcoin ci yml

kayabaNerve committed:
Multiple fixes to the Bitcoin processor

Doesn't unwrap on RPC errors. Returns the expected connection error.

Fee calculation has a random - 1. This has been removed.

Supports the change address being an Option, as it is. This should not have
been blindly unwrapped.

kayabaNerve committed:
Remove unnecessary RPC code

kayabaNerve committed:
Further RPC simplifications

kayabaNerve committed:
Simplify Bitcoin action

It should not be mining.

kayabaNerve committed:
cargo fmt

kayabaNerve committed:
Finish RPC simplifications

kayabaNerve committed:
Run bitcoind as a daemon

kayabaNerve committed:
Remove the requirement on txindex

Saves tens of GB.

Also has attempt_send no longer return a list of outputs. That's incompatible
with this and only relevant to old scheduling designs.

kayabaNerve committed:
Remove number from Bitcoin SignableTransaction

Monero requires the current block number for decoy selection. Bitcoin doesn't
have a use.

kayabaNerve committed:
Ban coinbase transactions

These are burdened by maturity, so it's critically flawed to support them.

This causes the test_send function to fail as its working was premised on
a coinbase output. While it does make an actual output, it had insufficient
funds for the test's expectations due to regtest halving every 150 blocks.

In order to workaround this, the test will invalidate any existing chain,
offering a fresh start.

Also removes test_get_spendables and simplifies test_send.

kayabaNerve committed:
Various simplifications

Modifies SpendableOutput further to not require RPC calls at time of sign.

Removes the need to have get_transaction in the RPC.

kayabaNerve committed:
Clean prepare_send

kayabaNerve committed:
Update the Bitcoin TransactionMachine to output a Transaction

kayabaNerve committed:
Bitcoin TransactionMachine simplifications

kayabaNerve committed:
Update XOnly key handling

kayabaNerve committed:
Use a single sighash cache

kayabaNerve committed:
Move tweak_keys

kayabaNerve committed:
Remove unnecessary PSBT sets

kayabaNerve committed:
Restore removed newlines

kayabaNerve committed:
Other newlines

kayabaNerve committed:
Replace calculate_weight's custom math with a dummy TX serialize

kayabaNerve committed:
Move BTC TX construction code from processor to bitcoin

kayabaNerve committed:
Rename transactions.rs to wallet.rs

kayabaNerve committed:
Remove unused crate

kayabaNerve committed:
Note TODO

kayabaNerve committed:
Clean bitcoin signature test

kayabaNerve committed:
Make unit test out of BTC FROST signing test

kayabaNerve committed:
Final lint

kayabaNerve committed:
Remove usage of PartiallySignedTransaction
